Code/Framework/AzCore/AzCore/Settings/ConfigParser.cpp

```diff
@@ -0,0 +1,211 @@
+            for (; startIndex < sectionName.size() && isspace(sectionName[startIndex]); ++startIndex);
+            sectionName = sectionName.substr(startIndex);
+            size_t endIndex = sectionName.size();
+            for (; endIndex > 0 && isspace(sectionName[endIndex - 1]); --endIndex);
+            sectionName = sectionName.substr(0, endIndex);
+        }
+
+        return sectionName;
+    }
+
+    auto ConfigParserSettings::DefaultDelimiterFunc(AZStd::string_view line) -> ConfigKeyValuePair
+    {
+        constexpr AZStd::string_view CommandLineArgumentDelimiters{ "=:" };
+        ConfigKeyValuePair keyValuePair;
+        keyValuePair.m_value = line;
+
+        // Splits the line on the first delimiter and stores that in the keyValuePair key variable
+        // The StringFunc::TokenizeNext function updates the key parameter in place
+        // to contain all the text after the first delimiter
+        // So if keyValuePair.m_key="foo = Hello Ice Cream=World:17", the call to TokenizeNext would
+        // split the value as follows
+        // keyValuePair.m_key = "foo"
+        // keyValuePair.m_value = "Hello Ice Cream=World:17"
+        if (auto keyOpt = AZ::StringFunc::TokenizeNext(keyValuePair.m_value, CommandLineArgumentDelimiters);
+            keyOpt.has_value())
+        {
+            // Strip the whitespace around the key part
+            keyValuePair.m_key = AZ::StringFunc::StripEnds(*keyOpt);
+        }
+
+        // Strip white spaces around the value part
+        keyValuePair.m_value = AZ::StringFunc::StripEnds(keyValuePair.m_value);
+        return keyValuePair;
+    };
+
+    ParseOutcome ParseConfigFile(AZ::IO::GenericStream& configStream, const ConfigParserSettings& configParserSettings)
+    {
+        // The trace system is not being used here as this function INI system is available
+        // before any allocators or initialization of any systems take place
+
+        ParseOutcome parseOutcome;
+        if (!configParserSettings.m_parseConfigEntryFunc)
+        {
+            return AZStd::unexpected(ParseErrorString::format("The Parse Config Entry function is not valid for parsing a Windows Style INI line"));
+        }
+
+        // The ConfigFile is parsed into a fixed_vector of ConfigBufferMaxSize below
+        // which avoids performing any dynamic memory allocations during parsing
+        // Only the user supplied callback functions as part of the ConfigParserSettings can potentially allocate memory
+        size_t remainingFileLength = configStream.GetLength();
+        if (remainingFileLength == 0)
+        {
+            return AZStd::unexpected(ParseErrorString::format(R"(Config stream "%s" is empty, nothing to merge)" "\n", configStream.GetFilename()));
+        }
+
+        constexpr size_t ConfigBufferMaxSize = 4096;
+        AZStd::fixed_vector<char, ConfigBufferMaxSize> configBuffer;
+        size_t readSize = (AZStd::min)(configBuffer.max_size(), remainingFileLength);
```

Comment:
maybe throw a warning if `configBuffer` is larger than the max size to let the user known that some parts of the stream won't be read.

Reply:
We can't actually throw a warning here.
This code is meant to be used in early startup before the allocators and trace system are available for reading startup files.

That is why this function returns an Outcome.

The preconditions of this function is that a single line cannot be larger than 4096 and that is documented in the function [declaration](https://github.com/o3de/o3de/pull/12957/files/66f05b8be6c8614771eed9531901d53f0b270a59#diff-ccedb6a20bd3fcb1df6f384991b4c3046b1b41d1ddc4fac09edbeb24fffa74f3R129-R130).

It is fine if the config file is larger than 4096 because it acts like a sliding 4096 byte window